### ISSUE TITLE
Wrong Logistic Loss

### ISSUE THREAD

**Issue by yk** [tensorflow/tensorflow#172]
In tensorflow/python/ops/nn.py, the logistic loss is described as

x - x * z + log(1 + exp(-x))

but this is not what's commonly known as the logistic loss.
I see that you're trying to apply a trick here to avoid overflow, but it should be 

x - x * z + log(exp(0 - (x - x * z)) + exp(-x))

instead of what you have.

Of course, the correct version is now prone to underflow, so it would make most sense to add a native robust log-sum-exp function to tensorflow and then rely on that to avoid over- and underflow.

**Comment by ludimagister** [comment]
Just to clarify what we do here:

Sigmoid input is x, output is y, target is z:

y = 1 / (1 + exp(-x))
1 - y = exp(-x) / (1 + exp(-x))

Logistic loss 
= -[z * log(y) + (1-z) * log(1-y)]
= z * log(1 + exp(-x)) - (1-z) * [-x - log(1 + exp(-x))]
= z * log(1 + exp(-x)) + x + log(1 + exp(-x)) - z*x - z * log(1 + exp(-x))
= x - z*x + log(1 + exp(-x))

To avoid overflow and branching in TensorFlow this is calculated as:

LogisticLoss = max(x, 0) - z*x + log(1 + exp(-abs(x)))

**Comment by jey-ivanov** [comment]
What is the purpose of replacing -x with -abs(x)?

**Comment by tngan** [comment]
@jey-ivanov 0 < exp(-abs(x)) <= 1 to avoid overflow. Without abs, the smaller the x value (negative), the greater the value of exp(-x).

**Comment by yk** [comment]
I see, you're assuming your targets to be 0/1, where as usually, people are using -1/+1 labels when speaking of the logistic loss.
Thanks for clarifying! Using that, I'm able to achieve the same test errors as with my naive log-sum-exp implementation.

Are you planning to add a robust log-sum-exp in the future?

**Comment by tianyang-li** [comment]
I think it would be nice for Tensor Flow to have a native log-sum-exp, probably that'll make things faster, and help people have less numerical issues.

**Comment by yaroslavvb** [comment]
It would be useful. Here's a similar CL which creates squared difference of two tensors. It's kind of complicated looking because same implementation works both on CPU and GPU: https://github.com/tensorflow/tensorflow/commit/bf536bcc888768b586440579b3fdeecf80a48acf

**Comment by vladfi1** [comment]
I have a decent log-sum-exp implemented:

    def log_sum_exp(xs):
      maxes = tf.reduce_max(xs, keep_dims=True)
      xs -= maxes
      return tf.squeeze(maxes, [-1]) + tf.log(tf.reduce_sum(tf.exp(xs), -1))

I think (?) this should avoid any major numerical issues...
